### ISSUE TITLE
Allow the user to input consumption/emission in other measurement units

### ISSUE THREAD

**Issue by Ceilican** [OiWorld/CarbonFootprintGoogleMaps#3]
It should be possible for the user to input the car's fuel consumption (in litres per km, or kilometres per litre or miles per gallon) and the extension would calculate the CO2 emissions (in g/km) automatically.

Code for such calculations can be found here: http://www.oiworld.org/map

Update: The link above is broken. See code below.

**Comment by ExC0d3** [comment]
Hi, I tried opening the link o see the code for the calculations, but got an Application Error, could you please confirm whether the link is active or not ?

**Comment by Ceilican** [comment]
Sorry for the broken link. I will post the code I had in mind below:

```
// Copyright © 2013 OiWorld. All rights reserved.


var fuels = {
  diesel : 0,
  gasoline : 1,
  lpg : 2,
  e10 : 3,
  e25 : 4,
  e85 : 5,
  ethanol : 6,
  biodiesel : 7
};

// Static Info
// ===========
//
// Source: http://www.biomassenergycentre.org.uk/portal/page?_pageid=75,163182&_dad=portal&_schema=PORTAL


// Energy Density, in MJ/l
var energyDensity = {}; 
energyDensity[fuels.diesel] = 36;
energyDensity[fuels.gasoline] = 32;
energyDensity[fuels.lpg] = 24;
energyDensity[fuels.e10] = (0.10 * 21) + (0.90 * 32) ;
energyDensity[fuels.e25] = (0.25 * 21) + (0.75 * 32);
energyDensity[fuels.e85] = (0.85 * 21) + (0.15 * 32);
energyDensity[fuels.ethanol] = 21;
energyDensity[fuels.biodiesel] = 33;

// CO2 emission on combustion, in g/l
var CO2Emission = {}; 
CO2Emission[fuels.diesel] = 2614;
CO2Emission[fuels.gasoline] = 2328;
CO2Emission[fuels.lpg] = 1533;
CO2Emission[fuels.e10] = (0.10 * 1503) + (0.90 * 2328) ;
CO2Emission[fuels.e25] = (0.25 * 1503) + (0.75 * 2328);
CO2Emission[fuels.e85] = (0.85 * 1503) + (0.15 * 2328);
CO2Emission[fuels.ethanol] = 1503;
CO2Emission[fuels.biodiesel] = 2486;

// approximate life-cycle GHG emission, in CO2 eq. g/l
var GHGEmission = {}; 
GHGEmission[fuels.diesel] = 3128;
GHGEmission[fuels.gasoline] = 2600;
GHGEmission[fuels.lpg] = 1533 * (2600/2328);    //assuming it grows as for gasoline
GHGEmission[fuels.e10] = (0.10 * (724 + 511)/2) + (0.90 * 2328) ;
GHGEmission[fuels.e25] = (0.25 * (724 + 511)/2) + (0.75 * 2328);
GHGEmission[fuels.e85] = (0.85 * (724 + 511)/2) + (0.15 * 2328);
GHGEmission[fuels.ethanol] = (724 + 511)/2;     // wheat = 511, sugar beet = 724 
GHGEmission[fuels.biodiesel] = (1334 + 437/2);  // waste vegetable oil = 437 , rapeseed oil = 1334


// Initialization
// ==============

// State variables
var fuel = 1;
var fuelEfficiency = 9;
var isUnitSystemImperial = false;


function initialize(f, fE, uS) {
  fuel = f;
  fuelEfficiency = fE;
  isUnitSystemImperial = uS;
  initializeMap();
  updateUI();
}

var directionsDisplay = new google.maps.DirectionsRenderer({draggable: true});
var directionsService = new google.maps.DirectionsService();


function initializeMap() {
  var worldCenter = new google.maps.LatLng(0, 0);
  var mapOptions = {
    zoom:2,
    mapTypeId: google.maps.MapTypeId.ROADMAP,
    center: worldCenter
  }
  var map = new google.maps.Map(document.getElementById('map_canvas'), mapOptions);
  directionsDisplay.setMap(map);
	
  google.maps.event.addListener(directionsDisplay, 'directions_changed', function() {
   updateUI();
  });
}



// Functions called by buttons and events of the UI
// ================================================

function calcRoute() {
  var start = document.getElementById('start').value;
  var end = document.getElementById('end').value;
  var request = {
	origin:start,
	destination:end,
	travelMode: google.maps.DirectionsTravelMode.DRIVING
  };
  directionsService.route(request, function(response, status) {
    if (status == google.maps.DirectionsStatus.OK) {
	  directionsDisplay.setDirections(response);
	  updateUI();
    }
  });
}


function keyPress(event) {
  if (event.keyCode == 13) calcRoute();
} 


function offset() {
  var emissionGHG = computeEmissions(getDistance(directionsDisplay.directions.routes[0])).GHG;
  var trees = Math.round(emissionGHG/8200); // Trees necessary for absorbing the emissions in one year
  
  window.location.href = "http://maps.oiworld.org/plant?trees=" + trees;
}
  
  
function changeFuelEfficiency() {
  fuelEfficiency = document.getElementById('fuel_efficiency_input').value; 
  updateUI();
}  
  
  
function changeUnitSystem() {
  isUnitSystemImperial = !isUnitSystemImperial;
  if (isUnitSystemImperial) fuelEfficiency = Math.round(fuelEfficiency * 2352.15)/1000;
  else fuelEfficiency = Math.round((fuelEfficiency * 1000) / 2.35215)/1000;   
  updateUI();
} 


function changeFuel() {
  var previousFuel = fuel;
  
  if (fuel < (fuelNames.length - 1)) fuel = fuel + 1;
  else fuel = 0;
  
  fuelEfficiency = fuelEfficiency * (energyDensity[fuel] / energyDensity[previousFuel]);
  
  updateUI();
} 
   


// Update of the UI
// ================  
  
function updateUI() {
  document.getElementById("unit_system_button").innerHTML = unitSystemToString();
  document.getElementById("fuel_type_button").innerHTML = fuelNames[fuel];
  document.getElementById('fuel_efficiency_input').value = Math.round(fuelEfficiency*10)/10;

  if (directionsDisplay.directions) {
    var distance = getDistance(directionsDisplay.directions.routes[0]);
    document.getElementById("distance").innerHTML = distanceToString(distance);
	
    var duration = getDuration(directionsDisplay.directions.routes[0]);
    document.getElementById("duration").innerHTML = durationToString(duration);
	
    var emissions = computeEmissions(distance);	
    document.getElementById("CO2").innerHTML = emissionToString(emissions.CO2, "CO<sub>2</sub>");
    document.getElementById("GHG").innerHTML = emissionToString(emissions.GHG, "CO<sub>2</sub>e <sub>(lifecycle)</sub>");
  
    document.getElementById("offset_button").style.visibility = "visible";
  }
}


function unitSystemToString() {
  return isUnitSystemImperial ? "mpg" : "km/l";
}   
  
   
function emissionToString(amount, poluent) { // amount in grams
  if (isUnitSystemImperial) {
    if (amount < 50000) return Math.round(amount*2.2046/100)/10 + " lb of " + poluent;
    else return Math.round(amount*2.2046/1000) + " lb of " + poluent;
  }
  else {
    if (amount < 1000) return amount + " g of " + poluent;
    else if (amount < 50000) return Math.round(amount/100)/10 + " kg of " + poluent;
    else return Math.round(amount/1000) + " kg of " + poluent;
  }
}


function distanceToString(distance) {
  if (isUnitSystemImperial) {
    if (distance < 500) return Math.round(distance/0.9144) + " yd";
    else if (distance < 50000) return Math.round(distance/160.93)/10 + " mi";
    else return Math.round(distance/1609.3) + " mi";
  }
  else {
    if (distance < 1000) return distance + " m";
    else if (distance < 50000) return Math.round(distance/100)/10 + " km";
    else return Math.round(distance/1000) + " km";
  }
}
 
    
function durationToString(duration) {
  var hours = Math.floor(duration / 3600);
  var minutes = Math.round((duration - (hours * 3600))/60); 
  return ((hours == 0) ? "" : (hours + " h ")) + minutes + " min";
}  
  
  

// Functions to calculate data from the route
// ==========================================
  
function getDistance(route) {
  var distance = 0;
  for (i = 0; i < route.legs.length; i++) {
    distance += route.legs[i].distance.value;
  }
  return distance; // in meters
}


function getDuration(route) {
  var duration = 0;
  for (i = 0; i < route.legs.length; i++) {
    duration += route.legs[i].duration.value;
  }
  return duration; // in seconds
}


function computeFuelAmount(distance) {
  var efficiency = isUnitSystemImperial ? (fuelEfficiency / 2.35215) : fuelEfficiency; // in km/l
  return ((distance/1000)/efficiency); // in liters
}


function computeEmissions(distance) {
  var fuelAmount = computeFuelAmount(distance);
  return {
    CO2: fuelAmount * CO2Emission[fuel],
    GHG: fuelAmount * GHGEmission[fuel]
  }
}
  

```

**Comment by Ceilican** [comment]
This issue has been partially solved in pull request 13: https://github.com/OiWorld/CarbonFootprintGoogleMaps/pull/13.

However, there is a related issue that still needs to be addressed, and its ideal solution would probably require a refactoring of the solution in pull request 13. The related issue is the following:

If a user specifies the emissions in, let's say, pounds instead of grams or kilograms, then he will probably want to see the emissions in pounds as well.

**Comment by Ceilican** [comment]
@ExC0d3 , It is important that you introduce yourself in our mailing-list: https://groups.google.com/forum/#!forum/cse-at-tu-wien-gsoc-2016

**Comment by goelrohan6** [comment]
@Ceilican I have solved the following issue in pull request #29 please review it accordingly and suggest changes if required, I have also solved the issue for showing units in pounds if the user prefers it.
UI updated, and the background image can be updated just by putting in the source folder by naming it background.jpg.
